src/cargo/util/config/mod.rs

```diff
@@ -393,8 +394,13 @@ impl Config {
     }
 
     /// Gets a reference to the shell, e.g., for writing error messages.
-    pub fn shell(&self) -> RefMut<'_, Shell> {
-        self.shell.borrow_mut()
+    pub fn shell(&self) -> MutexGuard<'_, Shell> {
```

Comment:
Can the docstring here include some warnings about the restrictions the caller should be careful about? For example, this should not be called in the same thread if that thread already has a shell (otherwise it would deadlock).

Overall I'm a bit nervous about this since there won't be any compile time checks, and there are a *lot* of places that get a shell. I don't think I can review all of those call sites. I think the current code is probably ok, otherwise the RefCel would have panic'ed, but a panic is a lot better than a deadlock.

Reply:
That's a good call, and I see the issue with going `RefCell` to `Mutex` as well. So much so that I'd also love to not have to do that.

For now I have added a big deadlock notice.